clang/lib/Driver/Driver.cpp

```diff
@@ -1290,6 +1292,101 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
         Diag(clang::diag::warn_drv_empty_joined_argument)
             << SYCLTargetsValues->getAsString(C.getInputArgs());
     }
+  }
+  // If the user specified --offload-arch, deduce the offloading
+  // target triple(s) from the set of architecture(s).
+  // Create a toolchain for each valid triple.
```

Comment:
Maybe add some additional comment about why `IsHIP` and `IsCuda` is not supported here.

Reply:
I don't think we have checked `cu` or `hip` types from the list of inputs previously anyways. My understanding is that this check here is is safe-guarding from the fact that a `HIP`/`Cuda` program may end up getting both `HIP`/`Cuda` offloading and `SYCL` offloading at the same time which we don't support but that shouldn't be possible to happen in `-fsycl` mode, right?
Regardless, I agree with @mdtoguchi that we need a comment for this, please, potentially test too but it's fine as is to me as well. It can also be abstracted, if you wish, in a static helper that returns if the input type is valid for SYCL offload which will be more self explanatory from the name of the function.

Reply:
> A potential test could be something along the lines: checking that `clang++ -fsycl ...` doesn't try to do SYCL offloading in `.cu` (for cuda type) and `.hip` (for hip type) files (there is already a `hip-offload-arch.hip` test file in the clang/test/Driver folder).

When you say `doesn't try to do SYCL offloading...` , do you mean that `-fsycl-is-device` option is not passed?

Also, I get the following error for the below invocation:
```
clang -### --driver-mode=g++ -nocudainc -nocudalib -fsycl --offload-new-driver --offload-arch=sm_80 sycl-offload-arch-cuda-type.cu
error: unable to create target: 'No available targets are compatible with triple "nvptx64-nvidia-cuda"'
```

Reply:
I tried the branch and couldn't see the above diagnostic error, so I am not sure about it. To your question, I think you are right, I haven't thought of what `doesn't try to do SYCL offloading` would look like too well. I guess a suggestion would be to essentially have a `not clang++ -fsycl --offload-new-driver --offload-arch` test run command for the `.cu` and `.hip` file types and a Filecheck `CHECK` for a diagnostic error string (using `not` so it doesn't fail the lit test as the clang invocation fails`) that should be emitted in those cases. To do so we'll probably need to move the `IsHIP && IsCuda` check inside the body of the `else if` and emit the Driver error diagnostic there. This is just a suggestion wrt testing the logic for that codepath, but as I said I am okay to leave it as is and added at a later point. For what it's worth it'll probably be appreciated for upstreaming I guess.

Reply:
AFAIC, for CUDA type files, we can do the following invocation:
`clang++  -fsycl -fsycl-targets=nvptx64-nvidia-cuda  -Xsycl-target-backend  --cuda-gpu-arch=sm_80   --cuda-gpu-arch=sm_80  -c  syclfile.cu`

Note that `--cuda-gpu-arch` is an alias for `--offload-arch`

However, when we do :

`clang++ -fsycl --offload-new-driver  --offload-arch=sm_80  syclFile.cu`

We get the following error:
`clang: error: unsupported option '--offload-new-driver'`

I believe this is because there is no SYCL offloading support for CUDA type files(.cu) when `--offload-new-driver` option is passed.

Reply:
Right, well let's leave it like this for now then, thanks! We can add further testing if needed later, also the mixing of cu and sycl device code is not really working (used to be experimental ages ago), so we can deal with it at a later point.